[hop] support local_map + SAC

xmfan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #163602
* __->__ #163322

Some ops like local_map hop's deferred mode are not desugared by make_fx, this means that when we apply SAC tags, we will need to define dispatch rules for the SAC torch dispatch modes as pointed out here: https://github.com/pytorch/pytorch/issues/162246#issuecomment-3259176721. This PR adds those rules.

Additionally it fixes a pre-existing issue where we weren't coercing tangent layout (that AOTAutograd typically does) when partitioning the HOP joint.

xmfan committed:
[hop] support local_map + SAC

[ghstack-poisoned]

xmfan committed:
Update on "[hop] support local_map + SAC"




[ghstack-poisoned]

xmfan commented:
hold on, updating test to use aot eager backend

soulitzer commented on test/higher_order_ops/test_local_map.py:
```diff
@@ -42,7 +44,17 @@ def context_parallel_attention(query, key, value):
+        op == torch.ops.aten._scaled_dot_product_flash_attention.default
+        or op == torch.ops.aten._scaled_dot_product_efficient_attention.default
+    ):
+        # NOTE: we can't save nondeterministic_seeded ops, the run with rng wrapper is not traceable yet
```
Wait do you mean "cannot recompute RNG ops"

xmfan replied:
oh ignore this, this is autoparallel frontend specific

xmfan committed:
Update on "[hop] support local_map + SAC"




[ghstack-poisoned]

xmfan committed:
Update on "[hop] support local_map + SAC"




[ghstack-poisoned]

ezyang reviewed (approved):
The actual impl changes are plausible

xmfan committed:
Update on "[hop] support local_map + SAC"




[ghstack-poisoned]

xmfan committed:
Update on "[hop] support local_map + SAC"


Some ops like local_map hop's deferred mode are not desugared by make_fx, this means that when we apply SAC tags, we will need to define dispatch rules for the SAC torch dispatch modes as pointed out here: https://github.com/pytorch/pytorch/issues/162246#issuecomment-3259176721. This PR adds those rules.

Additionally it fixes a pre-existing issue where we weren't coercing tangent layout (that AOTAutograd typically does) when partitioning the HOP joint.

[ghstack-poisoned]

xmfan commented:
@pytorchbot merge

ezyang commented on test/higher_order_ops/test_local_map.py:
```diff
@@ -193,10 +238,183 @@ def forward(self, q_1: "f32[8, 16, 16, 6]", k_1: "f32[8, 16, 16, 6]", v_1: "f32[
+    @requires_cuda_and_triton
+    @unittest.skipIf(
+        not torch.distributed.is_available(), "Torch distributed not available."
+    )
```
oh ok, please look forward to https://www.internalfb.com/diff/D82283623
